### DISCUSSION TITLE
randomise trials every block

### DISCUSSION THREAD

**Discussion by Jonas-Crucke** [jspsych/jsPsych#1190] [Q&A] [answered]
Hi, so I'm currently programming on my first experiment in jsPsych. It's a stroop task, but I'm trying to find whether people can use congruency of the previous trial to predict the color of the stroop stimulus on the current trial. So at the start of the experiment 2 colors are attributed to congruent trials, the other 2 to incongruent trials. The 2 attributed colors will both have a 40% chance to follow on a congruent or incongruent trial, depending on which condition they were assigned to. The other 2 colors will thus have a 10% chance of following a congruent or incongruent trial, depending on the condition. 

So for example: if red and blue are attributed to congruent, they will both have a 40% chance of being the color of the stimuli on the current trial, when the previous trial was a congruent trial. Green and Yellow will both have a 10% chance of being the stimulus color, when the previous trial was a congruent trial. Green and yellow will be attributed to incongruent, they will both have a chance of 40% of being the stimulus color after an incongruent trial, red and blue will both have a 10% chance.

This all happens in a seperate JS file which is then loaded in as a function in my html file.

This works, however, the trials are all in the same order across all blocks. I want them to be randomized, but the stimuli should stay the same (so if red and blue are attributed to congruent, they should be so for the whole experiment). I tried a number of things, such as shuffling the list on finish of the block, but this doesn't work.
Anybody an idea on how to fix this?

thanks!

**Comment by becky-gilbert** [comment]
Hi @Jonas-Crucke! Would you mind sharing the code you're using to set up the experiment, i.e. what your stimulus/trial information looks like, and how you're setting up the trials? You could share all of your code, or if it's very long and complicated, then maybe just a simplified example. Here's how to [create nicely-formatted code blocks in the comments](https://github.com/jspsych/jsPsych/discussions/1113), in case it's helpful.

**Reply by Jonas-Crucke** [reply]
Hi @becky-gilbert, thanks for your response!
So basically I have two files, in one file I first make 2 arrays:
````javascript
var congruent_color_array = jsPsych.randomization.shuffle(['red', 'blue', 'green', 'yellow']);
var incongruent_color_array = jsPsych.randomization.shuffle(['red', 'blue', 'green', 'yellow']);


while (congruent_color_array[0] == incongruent_color_array[0] || congruent_color_array[0] == incongruent_color_array[1] || congruent_color_array[1] == incongruent_color_array[0] || congruent_color_array[1] == incongruent_color_array[1] ){
  incongruent_color_array = jsPsych.randomization.shuffle(incongruent_color_array);
};

var equal_color = [19,19,5,5]; //this repeats the first 2 elements 19 times and the last 2 elements 5 times
var congruent_color_array = jsPsych.randomization.repeat(congruent_color_array, equal_color, false);
var incongruent_color_array = jsPsych.randomization.repeat(incongruent_color_array, equal_color,false);
````
Afterwards I write a function to which I refer in my actual experiment file. In this function I basically start with an empty trial list. which I then fill in at random, starting with congruency. Based on congruency, the color of the stimulus and the word is determined.
This function then returns a trial list.

````javascript
  var the_current_trial_list = make_trial_list();
````
the current trial list is then the timeline variable of the test procedure in my timeline.
I then use this timeline variable to call back the color and word to present in the experiment:

````javascript
var test = {
    type: 'html-keyboard-response',
    stimulus: function(){
      return '<p style="color: ' +jsPsych.timelineVariable('color', true)+'; font-size:32pt">'+jsPsych.timelineVariable('word', true)+'</p>';
    },
    choices: ['d','f','j','k'],
    on_finish: function(data){
      data.correct = jsPsych.pluginAPI.compareKeys(data.key_press, jsPsych.timelineVariable('correct_response', true));
      if (data.correct){
        data.correct = 1
      } else{
        data.correct = 0
      }
      data.block = blockNr
    }
  };
````
The problem is that the make_trial_list function is only executed once and not every block. With each block the function has to be repeated to overwrite the_current_trial_list. I tried using a function as a timeline variable but this gives errors. I hope this is clear enough for you?
Thanks in advance!

**Comment by becky-gilbert** [comment, marked as answer]
Hi @Jonas-Crucke, really sorry for the slow response! I'm not sure if you've solved this problem already, but thought I'd try to help anyway. 

I think you're right that you can't use a function to create different `timeline_variables` arrays during the experiment. If you just need the same stimuli in a randomized order, I'm wondering if you could shuffle the trial order for each block before the experiment starts, and use these different shuffled arrays across blocks? 

And for the block number, you could add the appropriate block number value to each trial list array (e.g. `{color: "red", word: "green", block_number: 1}`), and then add it to the `test` trial data with `data: {block_number: jsPsych.timelineVariable('block_number')}`. Or you could use a global variable to store the current block number, save it to the trial data using a dynamic parameter, and change the value of this variable during the task, e.g.
```javascript
var trial_list_1 = make_trial_list();
var trial_list_2 = jsPsych.randomization.shuffle(trial_list_1);

var curr_block_number = 1;

var test = {
  type: 'html-keyboard-response',
  stimulus: function(){
    return '<p style="color: ' +jsPsych.timelineVariable('color', true)+'; font-size:32pt">'+jsPsych.timelineVariable('word', true)+'</p>';
  },
  choices: ['d','f','j','k'],
  on_finish: function(data){
    data.correct = jsPsych.pluginAPI.compareKeys(data.key_press, jsPsych.timelineVariable('correct_response', true));
    if (data.correct){
      data.correct = 1
    } else{
      data.correct = 0
    }
  },
  data: function() {
    return {block: curr_block_number};
  }
};

var block_1 = {
  timeline: [test],
  timeline_variables: trial_list_1,
  on_finish: function() {
    curr_block_number = 2;
  }
}

var block_2 = {
  timeline: [test],
  timeline_variables: trial_list_2
}
```

**Reply by Jonas-Crucke** [reply]
Hi Becky,

I fixed it already, but did it in the same way as you suggest here! Thanks for your responses anyway :)
